File: demo_src_go/internal/store/store.go
// Package store provides a single-owner, checksummed, append-only JSON journal.
package store

import (
	"bufio"
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"golang.org/x/sys/unix"
	"io"
	"os"
	"path/filepath"
	d "secretary_go_demo/internal/domain"
	"sort"
	"strings"
	"sync"
)

type Store struct {
	transactionIDs   map[int64]string
	mu               sync.Mutex
	Root             string
	file, lock       *os.File
	records          map[string]d.R
	events, receipts []d.R
	seq, epoch       int64
	digest           string
	max              int
	poisoned         bool
}
type Tx struct {
	s                *Store
	changed          map[string]d.R
	events, receipts []d.R
}
type Frame struct {
	Payload string `json:"payload_b64"`
	Hash    string `json:"sha256"`
}

func Open(root string, max int) (s *Store, err error) {
	if !filepath.IsAbs(root) {
		root, err = filepath.Abs(root)
		if err != nil {
			return
		}
	}
	if err = mkdirDurable(filepath.Join(root, "journal")); err != nil {
		return
	}
	s = &Store{Root: root, records: map[string]d.R{}, transactionIDs: map[int64]string{}, digest: strings.Repeat("0", 64), max: max, epoch: 1}
	s.lock, err = os.OpenFile(filepath.Join(root, "owner.lock"), os.O_CREATE|os.O_RDWR, 0600)
	if err != nil {
		return nil, err
	}
	if err = unix.Flock(int(s.lock.Fd()), unix.LOCK_EX|unix.LOCK_NB); err != nil {
		s.lock.Close()
		return nil, fmt.Errorf("owner already running: %w", err)
	}
	defer func() {
		if err != nil {
			s.Close()
		}
	}()
	s.file, err = os.OpenFile(filepath.Join(root, "journal", "000001.jsonl"), os.O_CREATE|os.O_RDWR, 0600)
	if err != nil {
		return
	}
	if err = SyncDir(filepath.Join(root, "journal")); err != nil {
		return
	}
	if err = s.replay(); err != nil {
		return
	}
	if err = validateView(s.records); err != nil {
		return
	}
	s.epoch++
	_, err = s.file.Seek(0, io.SeekEnd)
	return
}
func (s *Store) Close() error {
	if s == nil {
		return nil
	}
	if s.file != nil {
		s.file.Close()
	}
	if s.lock != nil {
		unix.Flock(int(s.lock.Fd()), unix.LOCK_UN)
		return s.lock.Close()
	}
	return nil
}
func SyncDir(path string) error {
	f, e := os.Open(path)
	if e != nil {
		return e
	}
	defer f.Close()
	return f.Sync()
}
func Atomic(path string, b []byte) error {
	if e := mkdirDurable(filepath.Dir(path)); e != nil {
		return e
	}
	f, e := os.CreateTemp(filepath.Dir(path), ".write-")
	if e != nil {
		return e
	}
	tmp := f.Name()
	defer os.Remove(tmp)
	if _, e = f.Write(b); e == nil {
		e = f.Sync()
	}
	if x := f.Close(); e == nil {
		e = x
	}
	if e != nil {
		return e
	}
	if e = os.Rename(tmp, path); e != nil {
		return e
	}
	return SyncDir(filepath.Dir(path))
}
func (s *Store) Put(b []byte, media string) (d.R, error) {
	h := d.Hash(b)
	p := "objects/" + h[:2] + "/" + h
	ref := d.R{"path": p, "sha256": h, "bytes": int64(len(b)), "media_type": media}
	full := filepath.Join(s.Root, p)
	if old, e := os.ReadFile(full); e == nil {
		if !bytes.Equal(old, b) {
			return nil, errors.New("object content corruption")
		}
		return ref, nil
	} else if !os.IsNotExist(e) {
		return nil, e
	}
	if e := Atomic(full, b); e != nil {
		return nil, e
	}
	if e := SyncDir(filepath.Join(s.Root, "objects")); e != nil {
		return nil, e
	}
	return ref, nil
}
func (s *Store) Read(ref d.R) ([]byte, error) {
	h := d.S(ref["sha256"])
	if len(h) != 64 || d.S(ref["path"]) != "objects/"+h[:2]+"/"+h {
		return nil, errors.New("invalid object path")
	}
	p := filepath.Join(s.Root, d.S(ref["path"]))
	for x := p; x != s.Root; x = filepath.Dir(x) {
		i, e := os.Lstat(x)
		if e != nil {
			return nil, e
		}
		if i.Mode()&os.ModeSymlink != 0 {
			return nil, errors.New("symlink object")
		}
	}
	b, e := os.ReadFile(p)
	if e != nil {
		return nil, e
	}
	if d.Hash(b) != h || int64(len(b)) != d.N(ref["bytes"]) {
		return nil, errors.New("object digest/length mismatch")
	}
	return b, nil
}
func (s *Store) CheckRefs(v any) error {
	switch x := v.(type) {
	case map[string]any:
		if _, ok := x["sha256"]; ok && x["path"] != nil && x["bytes"] != nil {
			_, e := s.Read(x)
			return e
		}
		for _, v := range x {
			if e := s.CheckRefs(v); e != nil {
				return e
			}
		}
	case []any:
		for _, v := range x {
			if e := s.CheckRefs(v); e != nil {
				return e
			}
		}
	}
	return nil
}
func key(typ, id string) string        { return typ + "/" + id }
func (s *Store) View(typ string) []d.R { s.mu.Lock(); defer s.mu.Unlock(); return s.view(typ) }
func (s *Store) view(typ string) []d.R {
	r := []d.R{}
	for _, v := range s.records {
		if d.S(v["record_type"]) == typ {
			r = append(r, d.Clone(v))
		}
	}
	sort.Slice(r, func(i, j int) bool { return d.S(r[i]["id"]) < d.S(r[j]["id"]) })
	return r
}
func (s *Store) Get(typ, id string) d.R {
	s.mu.Lock()
	defer s.mu.Unlock()
	if r := s.records[key(typ, id)]; r != nil {
		return d.Clone(r)
	}
	return nil
}
func (s *Store) Events() []d.R {
	s.mu.Lock()
	defer s.mu.Unlock()
	a := []d.R{}
	for _, r := range s.events {
		a = append(a, d.Clone(r))
	}
	return a
}
func (s *Store) Epoch() int64    { return s.epoch }
func (s *Store) Sequence() int64 { s.mu.Lock(); defer s.mu.Unlock(); return s.seq }
func (t *Tx) Get(typ, id string) d.R {
	if r := t.changed[key(typ, id)]; r != nil {
		return d.Clone(r)
	}
	if r := t.s.records[key(typ, id)]; r != nil {
		return d.Clone(r)
	}
	return nil
}
func (t *Tx) List(typ string) []d.R {
	m := map[string]d.R{}
	for _, r := range t.s.view(typ) {
		m[d.S(r["id"])] = r
	}
	for _, r := range t.changed {
		if d.S(r["record_type"]) == typ {
			m[d.S(r["id"])] = d.Clone(r)
		}
	}
	a := []d.R{}
	for _, r := range m {
		a = append(a, r)
	}
	return a
}
func (t *Tx) Save(r d.R) error {
	k := key(d.S(r["record_type"]), d.S(r["id"]))
	old := t.s.records[k]
	if immutableRecord(d.S(r["record_type"])) && (old != nil || t.changed[k] != nil) {
		return fmt.Errorf("immutable %s cannot be saved twice", r["record_type"])
	}
	next := d.Clone(r)
	if old != nil {
		if d.N(r["revision"]) != d.N(old["revision"]) {
			return errors.New("CONFLICT revision")
		}
		next["revision"] = d.N(old["revision"]) + 1
		if r["record_type"] == "Execution" && !d.Edge("Retention", d.S(old["retention_state"]), d.S(next["retention_state"])) {
			return errors.New("invalid retention transition")
		}
		if a, b := d.S(old["state"]), d.S(next["state"]); a != "" && b != "" && !d.Edge(d.Machine(d.S(r["record_type"])), a, b) {
			return fmt.Errorf("invalid state edge %s %s -> %s", r["record_type"], a, b)
		}
	}
	if r["record_type"] == "Session" {
		next["last_journal_seq"] = t.s.seq + 1
	}
	next["updated_at"] = d.Now()
	t.changed[k] = next
	return nil
}
func (t *Tx) Object(v any) (d.R, error) { return t.s.Put(d.Bytes(v), "application/json") }
func (t *Tx) Log(kind, actor string, scope d.R, corr string, payload any) (d.R, error) {
	ref, e := t.Object(payload)
	if e != nil {
		return nil, e
	}
	r := d.Empty("OperationLogRecord")
	r["event_id"] = d.ID()
	r["stream"] = "SYSTEM"
	if scope["execution_id"] != nil {
		r["stream"] = "TASK"
	} else if scope["session_id"] != nil {
		r["stream"] = "MAIN"
	}
	r["scope"] = scope
	r["sequence"] = int64(len(t.s.events) + len(t.events) + 1)
	r["event_type"] = kind
	r["occurred_at"] = d.Now()
	r["recorded_at"] = d.Now()
	r["actor"] = actor
	r["payload"] = ref
	r["correlation_id"] = corr
	t.events = append(t.events, r)
	return r, nil
}
func (t *Tx) Receipt(id, hash, obj string) (d.R, error) {
	r := d.Empty("CommandReceipt")
	r["request_id"] = id
	r["request_hash"] = hash
	r["status"] = "ACCEPTED"
	r["object_id"] = obj
	r["journal_seq"] = t.s.seq + 1
	t.receipts = append(t.receipts, r)
	return r, nil
}
func (t *Tx) Existing(id, hash string) (d.R, error) {
	for _, r := range t.s.receipts {
		if d.S(r["request_id"]) == id {
			if d.S(r["request_hash"]) != hash {
				return nil, errors.New("CONFLICT request_id reused with different bytes")
			}
			return d.Clone(r), nil
		}
	}
	return nil, nil
}
func (s *Store) Update(fn func(*Tx) error) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.poisoned {
		return errors.New("RECOVERY_BLOCKED writer requires reopen")
	}
	t := &Tx{s: s, changed: map[string]d.R{}}
	if e := fn(t); e != nil {
		return e
	}
	if len(t.changed)+len(t.events)+len(t.receipts) == 0 {
		return nil
	}
	candidate := map[string]d.R{}
	for k, v := range s.records {
		candidate[k] = v
	}
	for k, v := range t.changed {
		candidate[k] = v
	}
	if e := validateView(candidate); e != nil {
		return e
	}
	j := d.Empty("JournalTransaction")
	j["txn_id"] = d.ID()
	j["sequence"] = s.seq + 1
	j["owner_epoch"] = s.epoch
	j["previous_digest"] = s.digest
	j["committed_at"] = d.Now()
	mut := []any{}
	for k, r := range t.changed {
		if e := d.Validate(d.S(r["record_type"]), r); e != nil {
			return fmt.Errorf("%s: %w", k, e)
		}
		if e := s.CheckRefs(r); e != nil {
			return e
		}
		ref, e := t.Object(r)
		if e != nil {
			return e
		}
		var rev int64
		if old := s.records[k]; old != nil {
			rev = d.N(old["revision"])
		}
		mut = append(mut, d.R{"object_type": r["record_type"], "object_id": r["id"], "expected_revision": rev, "new_revision": rev + 1, "snapshot": ref})
	}
	j["mutations"] = mut
	j["log_records"] = t.events
	j["receipts"] = t.receipts
	if j["log_records"] == nil || t.events == nil {
		j["log_records"] = []any{}
	}
	if t.receipts == nil {
		j["receipts"] = []any{}
	}
	if e := d.Validate("JournalTransaction", j); e != nil {
		return e
	}
	b := d.Bytes(j)
	if len(b) > s.max {
		return errors.New("journal frame too large")
	}
	f := Frame{base64.StdEncoding.EncodeToString(b), d.Hash(b)}
	line := append(d.Bytes(f), '\n')
	if _, e := s.file.Write(line); e != nil {
		s.poisoned = true
		return e
	}
	if e := s.file.Sync(); e != nil {
		s.poisoned = true
		return e
	}
	for k, r := range t.changed {
		s.records[k] = r
	}
	s.events = append(s.events, t.events...)
	s.receipts = append(s.receipts, t.receipts...)
	s.seq++
	s.transactionIDs[s.seq] = d.S(j["txn_id"])
	s.digest = f.Hash
	return nil
}
func (s *Store) replay() error {
	r := bufio.NewReaderSize(s.file, 65536)
	var offset int64
	for {
		line, e := r.ReadBytes('\n')
		if e == io.EOF {
			if len(line) > 0 {
				if x := Atomic(filepath.Join(s.Root, "journal", "incomplete-tail.bin"), line); x != nil {
					return x
				}
				if x := s.file.Truncate(offset); x != nil {
					return x
				}
				if x := s.file.Sync(); x != nil {
					return x
				}
			}
			break
		}
		if e != nil {
			return e
		}
		if len(line) > s.max*2 {
			return errors.New("RECOVERY_BLOCKED oversized frame")
		}
		var f Frame
		if e = d.Decode(line, &f); e != nil {
			return fmt.Errorf("RECOVERY_BLOCKED frame: %w", e)
		}
		b, e := base64.StdEncoding.DecodeString(f.Payload)
		if e != nil || len(b) > s.max || d.Hash(b) != f.Hash {
			return errors.New("RECOVERY_BLOCKED checksum")
		}
		var j d.R
		if e = d.Decode(b, &j); e != nil {
			return e
		}
		if e = d.Validate("JournalTransaction", j); e != nil {
			return e
		}
		if d.N(j["sequence"]) != s.seq+1 || d.S(j["previous_digest"]) != s.digest {
			return errors.New("RECOVERY_BLOCKED chain")
		}
		for _, v := range d.A(j["mutations"]) {
			m := d.M(v)
			b, e = s.Read(d.M(m["snapshot"]))
			if e != nil {
				return fmt.Errorf("RECOVERY_BLOCKED snapshot: %w", e)
			}
			var obj d.R
			if e = json.Unmarshal(b, &obj); e != nil {
				return e
			}
			if e = d.Validate(d.S(m["object_type"]), obj); e != nil {
				return e
			}
			k := key(d.S(m["object_type"]), d.S(m["object_id"]))
			rev := int64(0)
			if old := s.records[k]; old != nil {
				if immutableRecord(d.S(m["object_type"])) {
					return fmt.Errorf("RECOVERY_BLOCKED immutable %s modified", m["object_type"])
				}
				rev = d.N(old["revision"])
			}
			if d.N(m["expected_revision"]) != rev || d.N(m["new_revision"]) != rev+1 || d.N(obj["revision"]) != rev+1 || obj["id"] != m["object_id"] {
				return errors.New("RECOVERY_BLOCKED revision")
			}
			if e = s.CheckRefs(obj); e != nil {
				return fmt.Errorf("RECOVERY_BLOCKED reference: %w", e)
			}
			s.records[k] = obj
		}
		for _, v := range d.A(j["log_records"]) {
			ev := d.M(v)
			if d.N(ev["sequence"]) != int64(len(s.events)+1) {
				return errors.New("RECOVERY_BLOCKED event sequence")
			}
			if e = s.CheckRefs(ev); e != nil {
				return e
			}
			s.events = append(s.events, ev)
		}
		for _, v := range d.A(j["receipts"]) {
			s.receipts = append(s.receipts, d.M(v))
		}
		s.seq++
		s.transactionIDs[s.seq] = d.S(j["txn_id"])
		s.epoch = d.N(j["owner_epoch"])
		s.digest = f.Hash
		offset += int64(len(line))
	}
	return nil
}
func (s *Store) EventsUnsafe(t *Tx) []d.R {
	if t.s != s {
		panic("wrong transaction")
	}
	return s.events
}

func (s *Store) TransactionForRequest(id string) string {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, r := range s.receipts {
		if r["request_id"] == id {
			return s.transactionIDs[d.N(r["journal_seq"])]
		}
	}
	return ""
}
func (s *Store) KnownRef(ref d.R) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	wanted := d.Hash(d.Bytes(ref))
	var contains func(any) bool
	contains = func(v any) bool {
		switch x := v.(type) {
		case map[string]any:
			if x["sha256"] != nil && x["path"] != nil && d.Hash(d.Bytes(x)) == wanted {
				return true
			}
			for _, v := range x {
				if contains(v) {
					return true
				}
			}
		case []any:
			for _, v := range x {
				if contains(v) {
					return true
				}
			}
		}
		return false
	}
	for _, r := range s.records {
		if contains(r) {
			return true
		}
	}
	for _, r := range s.events {
		if contains(r) {
			return true
		}
	}
	return false
}

File: demo_pi/pi_secretary/src/store.ts
import { createHash, randomUUID } from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { Ajv2020 } from "ajv/dist/2020.js";
import addFormats from "ajv-formats";
import type {
  Contract,
  ObjectRef,
  OperationLogRecord,
  JournalTransaction,
} from "./contracts.ts";

export const root = fileURLToPath(new URL("../../../", import.meta.url));
export const id = (): string => randomUUID();
export const now = () => new Date().toISOString();
export const hash = (data: string | Buffer) =>
  createHash("sha256").update(data).digest("hex");
export const base = (recordID = id()) => ({
  id: recordID,
  revision: 1,
  updated_at: now(),
});
export function revise<T extends { revision: number; updated_at: string }>(
  r: T,
  patch: Partial<T>,
): T {
  return { ...r, ...patch, revision: r.revision + 1, updated_at: now() };
}
export type Stored = Extract<Contract, { id: string; revision: number }>;
const schema = JSON.parse(
  fs.readFileSync(path.join(root, "json/contracts.schema.json"), "utf8"),
);
const ajv = new Ajv2020({ strict: false, allErrors: true });
(addFormats as unknown as (instance: Ajv2020) => void)(ajv);
const validate = ajv.compile(schema);
export function shapeDefinition(name: string, value: unknown) {
  const validator = ajv.getSchema("urn:secretary:contracts:v1#/$defs/" + name);
  if (!validator || !validator(value))
    throw Error("INVALID_" + name + ": " + JSON.stringify(validator?.errors));
}
export function shape(record: unknown): asserts record is Contract {
  if (!validate(record)) {
    const kind = (record as { record_type?: string })?.record_type;
    const specific = kind
      ? ajv.getSchema("urn:secretary:contracts:v1#/$defs/" + kind)
      : undefined;
    if (specific) specific(record);
    throw Error(
      "INVALID_CONTRACT: " +
        JSON.stringify((specific?.errors ?? validate.errors)?.slice(0, 5)),
    );
  }
}
const machines = JSON.parse(
  fs.readFileSync(path.join(root, "state_machine/catalog.json"), "utf8"),
).machines as Record<
  string,
  { transitions: { from: string; event: string; to: string }[] }
>;
export function next(machine: string, state: string, event: string): string {
  const edge = machines[machine]?.transitions.find(
    (t) => t.from === state && t.event === event,
  );
  if (!edge) throw Error(`INVALID_TRANSITION ${machine} ${state} ${event}`);
  return edge.to;
}
function syncDir(dir: string) {
  const fd = fs.openSync(dir, "r");
  try {
    fs.fsyncSync(fd);
  } finally {
    fs.closeSync(fd);
  }
}
export class Store {
  readonly dir: string;
  epoch = 1;
  sequence = 0;
  digest = "0".repeat(64);
  eventSequence = 0;
  private records = new Map<string, Stored>();
  readonly logs: OperationLogRecord[] = [];
  private helper: ChildProcessWithoutNullStreams;
  private closed = false;
  private healthy = true;
  private eventTransactions = new Map<string, string>();
  transactionForEvent(eventID: string) {
    const tx = this.eventTransactions.get(eventID);
    if (!tx) throw Error("EVENT_NOT_COMMITTED");
    return tx;
  }
  private receipts = new Map<string, { hash: string; value: unknown }>();
  private constructor(dir: string, helper: ChildProcessWithoutNullStreams) {
    this.dir = dir;
    this.helper = helper;
    helper.once("exit", () => {
      this.healthy = false;
    });
  }
  static async open(dir: string) {
    fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
    const helper = spawn(
      "python3",
      [
        fileURLToPath(new URL("./lock.py", import.meta.url)),
        path.join(dir, "owner.lock"),
      ],
      { stdio: "pipe" },
    );
    await new Promise<void>((resolve, reject) => {
      helper.once("error", reject);
      helper.stdout.once("data", (b) =>
        String(b).trim() === "LOCKED" ? resolve() : reject(Error("OWNER_BUSY")),
      );
      helper.once("exit", (code) => {
        if (code) reject(Error("OWNER_BUSY"));
      });
    });
    const s = new Store(dir, helper);
    try {
      fs.mkdirSync(path.join(dir, "objects"), { recursive: true });
      s.replay();
      s.epoch =
        s
          .all()
          .reduce(
            (n, r) =>
              Math.max(n, Number("owner_epoch" in r ? r.owner_epoch : 0)),
            0,
          ) + 1;
      return s;
    } catch (e) {
      await s.close();
      throw e;
    }
  }
  all<T extends Stored = Stored>(type?: T["record_type"]): T[] {
    return structuredClone(
      [...this.records.values()].filter((r) => !type || r.record_type === type),
    ) as T[];
  }
  get<T extends Stored>(type: T["record_type"], recordID: string): T {
    const r = this.records.get(type + ":" + recordID);
    if (!r) throw Error("NOT_FOUND");
    return structuredClone(r) as T;
  }
  find<T extends Stored>(
    type: T["record_type"],
    recordID: string,
  ): T | undefined {
    try {
      return this.get<T>(type, recordID);
    } catch {
      return undefined;
    }
  }
  put(value: unknown, media = "application/json"): ObjectRef {
    const bytes = Buffer.isBuffer(value)
      ? value
      : Buffer.from(typeof value === "string" ? value : JSON.stringify(value));
    const sha = hash(bytes);
    const rel = "objects/" + sha;
    const file = path.join(this.dir, rel);
    if (!fs.existsSync(file)) {
      const temp = file + "." + id();
      const fd = fs.openSync(temp, "wx", 0o600);
      try {
        fs.writeFileSync(fd, bytes);
        fs.fsyncSync(fd);
      } finally {
        fs.closeSync(fd);
      }
      fs.renameSync(temp, file);
      syncDir(path.dirname(file));
    }
    const ref = {
      path: rel,
      sha256: sha,
      bytes: bytes.length,
      media_type: media,
    };
    this.bytes(ref);
    return ref;
  }
  bytes(ref: ObjectRef): Buffer {
    if (!/^objects\/[a-f0-9]{64}$/.test(ref.path))
      throw Error("INVALID_OBJECT_PATH");
    const f = path.join(this.dir, ref.path);
    if (fs.lstatSync(f).isSymbolicLink()) throw Error("OBJECT_SYMLINK");
    const b = fs.readFileSync(f);
    if (hash(b) !== ref.sha256 || b.length !== ref.bytes)
      throw Error("CORRUPT_OBJECT");
    return b;
  }
  read<T>(ref: ObjectRef): T {
    return JSON.parse(this.bytes(ref).toString()) as T;
  }
  receipt<T>(request: string, digest: string): T | undefined {
    const r = this.receipts.get(request);
    if (r && r.hash !== digest) throw Error("REQUEST_CONFLICT");
    return r?.value as T | undefined;
  }
  event(
    kind: string,
    payload: unknown,
    scope: OperationLogRecord["scope"] = {
      session_id: null,
      task_id: null,
      execution_id: null,
    },
    actor: OperationLogRecord["actor"] = "HOST",
  ): OperationLogRecord {
    return {
      record_type: "OperationLogRecord",
      schema_version: 1,
      event_id: id(),
      stream: scope.execution_id
        ? "TASK"
        : scope.session_id
          ? "MAIN"
          : "SYSTEM",
      scope,
      sequence: this.eventSequence + 1,
      event_type: kind,
      occurred_at: now(),
      recorded_at: now(),
      actor,
      payload: this.put(payload),
      causation_id: null,
      correlation_id: scope.execution_id ?? scope.session_id ?? id(),
      related_object_ids: [],
    };
  }
  commit(
    records: Stored[],
    events: OperationLogRecord[] = [],
    receipt?: { request: string; hash: string; value: string },
  ) {
    if (this.closed || !this.healthy || this.helper.exitCode !== null)
      throw Error("STORE_NOT_OWNER");
    for (const r of records) {
      shape(r);
      const old = this.records.get(r.record_type + ":" + r.id);
      if (r.revision !== (old?.revision ?? 0) + 1)
        throw Error("REVISION_CONFLICT");
    }
    if (
      new Set(records.map((r) => r.record_type + ":" + r.id)).size !==
      records.length
    )
      throw Error("DUPLICATE_MUTATION");
    const txn: JournalTransaction = {
      record_type: "JournalTransaction",
      schema_version: 1,
      txn_id: id(),
      sequence: this.sequence + 1,
      owner_epoch: this.epoch,
      previous_digest: this.digest,
      mutations: records.map((r) => ({
        object_type: r.record_type,
        object_id: r.id,
        expected_revision: r.revision - 1,
        new_revision: r.revision,
        snapshot: this.put(r),
      })),
      log_records: events.map((e, i) => ({
        ...e,
        sequence: this.eventSequence + i + 1,
      })),
      receipts: receipt
        ? [
            {
              record_type: "CommandReceipt",
              schema_version: 1,
              request_id: receipt.request,
              request_hash: receipt.hash,
              status: "ACCEPTED",
              object_id: receipt.value,
              reason: null,
              journal_seq: this.sequence + 1,
            },
          ]
        : [],
      committed_at: now(),
    };
    shape(txn);
    const staged = this.stage(txn);
    const payload = JSON.stringify(txn);
    const digest = hash(payload);
    const frame =
      JSON.stringify({
        payload_b64: Buffer.from(payload).toString("base64"),
        sha256: digest,
      }) + "\n";
    const file = path.join(this.dir, "journal.jsonl");
    const fd = fs.openSync(file, "a", 0o600);
    try {
      fs.writeFileSync(fd, frame);
      fs.fsyncSync(fd);
      syncDir(this.dir);
    } catch (e) {
      this.healthy = false;
      throw e;
    } finally {
      fs.closeSync(fd);
    }
    this.install(txn, digest, staged);
  }
  private stage(txn: JournalTransaction): Stored[] {
    const staged: Stored[] = [];
    const keys = new Set<string>();
    for (const m of txn.mutations) {
      const key = m.object_type + ":" + m.object_id;
      if (keys.has(key)) throw Error("DUPLICATE_MUTATION");
      keys.add(key);
      const old = this.records.get(key);
      if (
        (old?.revision ?? 0) !== m.expected_revision ||
        m.new_revision !== m.expected_revision + 1
      )
        throw Error("CORRUPT_REVISION");
      const r = this.read<Stored>(m.snapshot);
      shape(r);
      if (
        r.record_type !== m.object_type ||
        r.id !== m.object_id ||
        r.revision !== m.new_revision
      )
        throw Error("CORRUPT_SNAPSHOT");
      this.verifyRefs(r);
      if (
        old &&
        ["Context", "Checkpoint", "TaskResult", "ArchiveManifest"].includes(
          r.record_type,
        )
      )
        throw Error("IMMUTABLE_RECORD");
      if (old?.record_type === "Input" && r.record_type === "Input") {
        if (
          r.session_id !== old.session_id ||
          r.payload.sha256 !== old.payload.sha256
        )
          throw Error("INPUT_IDENTITY_CHANGED");
        if (
          old.state !== r.state &&
          !(
            (old.state === "ACCEPTED" && r.state === "CLAIMED") ||
            (old.state === "CLAIMED" && r.state === "HANDLED")
          )
        )
          throw Error("INVALID_INPUT_TRANSITION");
      }
      staged.push(r);
    }
    const candidate = new Map(this.records);
    for (const r of staged) candidate.set(r.record_type + ":" + r.id, r);
    if (
      [...candidate.values()].filter((r) => r.record_type === "Session")
        .length > 1
    )
      throw Error("MULTIPLE_MAIN_SESSIONS");
    for (const r of staged) {
      if (
        "session_id" in r &&
        r.session_id &&
        !candidate.has("Session:" + r.session_id)
      )
        throw Error("MISSING_SESSION");
    }
    const eventIDs = new Set<string>();
    for (const [index, e] of txn.log_records.entries()) {
      this.verifyRefs(e);
      if (
        e.sequence !== this.eventSequence + index + 1 ||
        eventIDs.has(e.event_id) ||
        this.eventTransactions.has(e.event_id)
      )
        throw Error("CORRUPT_EVENT_SEQUENCE");
      eventIDs.add(e.event_id);
    }
    const requests = new Set<string>();
    for (const r of txn.receipts) {
      if (requests.has(r.request_id) || this.receipts.has(r.request_id))
        throw Error("DUPLICATE_RECEIPT");
      requests.add(r.request_id);
    }
    return staged;
  }
  private install(txn: JournalTransaction, digest: string, staged: Stored[]) {
    for (const r of staged) this.records.set(r.record_type + ":" + r.id, r);
    for (const e of txn.log_records) {
      this.eventSequence = e.sequence;
      this.logs.push(e);
      this.eventTransactions.set(e.event_id, txn.txn_id);
    }
    for (const r of txn.receipts)
      this.receipts.set(r.request_id, {
        hash: r.request_hash,
        value: r.object_id,
      });
    this.sequence = txn.sequence;
    this.digest = digest;
  }
  private verifyRefs(value: unknown) {
    if (!value || typeof value !== "object") return;
    if (Array.isArray(value)) {
      for (const x of value) this.verifyRefs(x);
      return;
    }
    const v = value as Record<string, unknown>;
    if (
      typeof v.path === "string" &&
      typeof v.sha256 === "string" &&
      typeof v.bytes === "number" &&
      typeof v.media_type === "string"
    ) {
      this.bytes(v as unknown as ObjectRef);
      return;
    }
    for (const child of Object.values(v)) this.verifyRefs(child);
  }
  private replay() {
    const file = path.join(this.dir, "journal.jsonl");
    if (!fs.existsSync(file)) return;
    const data = fs.readFileSync(file);
    const last = data.lastIndexOf(10);
    if (last !== data.length - 1) {
      fs.writeFileSync(file + ".incomplete-" + id(), data.subarray(last + 1));
      fs.truncateSync(file, last + 1);
    }
    const text = data.subarray(0, last + 1).toString();
    for (const line of text.split("\n").slice(0, -1)) {
      const frame = JSON.parse(line);
      if (
        !frame ||
        Object.keys(frame).sort().join(",") !== "payload_b64,sha256" ||
        typeof frame.payload_b64 !== "string" ||
        typeof frame.sha256 !== "string"
      )
        throw Error("CORRUPT_FRAME");
      const payload = Buffer.from(frame.payload_b64, "base64");
      if (payload.toString("base64") !== frame.payload_b64)
        throw Error("CORRUPT_BASE64");
      if (hash(payload) !== frame.sha256) throw Error("CORRUPT_JOURNAL");
      const txn = JSON.parse(payload.toString()) as JournalTransaction;
      shape(txn);
      if (
        txn.sequence !== this.sequence + 1 ||
        txn.previous_digest !== this.digest
      )
        throw Error("CORRUPT_CHAIN");
      this.install(txn, frame.sha256, this.stage(txn));
    }
  }
  async close() {
    if (this.closed) return;
    this.closed = true;
    const exited = new Promise<void>((r) =>
      this.helper.once("exit", () => r()),
    );
    if (this.helper.exitCode !== null) return;
    this.helper.stdin.end();
    await exited;
  }
}
